rust/src/state_machine/phases/mod.rs

```diff
@@ -57,10 +60,19 @@ pub trait Phase {
     fn next(self) -> Option<StateMachine>;
 }
 
-/// A trait that must be implemented by a state to handle a request.
+/// A trait that must be implemented by a state to handle and reject a request.
 pub trait Handler {
     /// Handles a request.
     fn handle_request(&mut self, req: StateMachineRequest) -> Result<(), PetError>;
+
+    /// Rejects a request by returning a [`PetError::InvalidMessage`] back to the caller.
+    fn reject_request(&mut self) -> Result<(), PetError> {
```

Comment:
It feels a bit weird to put that here imo. Not sure but could we instead move that to the `process_single` method?

```rust
impl<S> PhaseState<S>
where
    Self: Handler + Phase,
{
    async fn process_single(&mut self) -> Result<(), StateError> {
        let (req, resp_tx) = self.next_request().await?;
        let span = req.span();
        let _span_guard = span.enter();
        let res = self.handle_request(req.into_inner());

        // if res is an InvalidMessage, do the `metrics!` stuff here
        // ...
        // ...

        // This may error out if the receiver has already be dropped but
        // it doesn't matter for us.
        let _ = resp_tx.send(res.map_err(Into::into));
        Ok(())
    }
}
```

Reply:
I like that idea but it would be an empty `if` if we compile it without the `metrics` feature. But the compiler will probably optimize that away. 
I will add the macro to the `process_single` method.

Reply:
> But the compiler will probably optimize that away.

For sure yeah.